### PR TITLE
dialects: cmpf operation in arith dialect

### PR DESCRIPTION
**Pull request by mesham**
Added cmpf operation in arith dialect, which is a comparison for floating point numbers. I have based this on the structure of the existing cmpi operation. Associated tests included.

### CODE REVIEW

**Comment by mesham** [comment]
Converted to draft, think I can improve this by unifying cmpi and cmpf, I'll do that tomorrow morning

File: xdsl/dialects/arith.py [hunk 1]
```diff
@@ -474,6 +474,95 @@ def from_mnemonic(operand1: Union[Operation, SSAValue],
...
+
+    @staticmethod
+    def _test_equality_of_operand_types(operand1: SSAValue,
+                                        operand2: SSAValue):
+        if operand1.typ == operand2.typ:
+          return True
+        return False
+
+    @staticmethod
+    def get(operand1: SSAValue | Operation,
```
Comment (math-fehr) [inline xdsl/dialects/arith.py:516]: Can you maybe merge this one with the next one? Like `arg : int | str`.

**Review by math-fehr — APPROVED** [review]
Nice! I would maybe merge the two constructors?

File: tests/dialects/test_arith.py [hunk 2]
```diff
@@ -82,3 +82,46 @@ def test_extend_truncate_fpops():
     assert ext_op.result.typ == f64
     assert trunc_op.input == b.result
     assert trunc_op.result.typ == f32
+
+
+def test_cmpf_from_mnemonic():
+    a = Constant.from_float_and_width(1.0, f64)
+    b = Constant.from_float_and_width(2.0, f64)
+    cmpi_ops = [None] * 10
```
Comment (webmiche) [inline tests/dialects/test_arith.py:90]:
```suggestion
    cmpf_ops = [None] * 10
```

File: tests/dialects/test_arith.py [hunk 3]
```diff
@@ -82,3 +82,46 @@ def test_extend_truncate_fpops():
...
+        assert op.lhs.typ == f64
+        assert op.rhs.typ == f64
+        assert op.predicate.value.data == index
+
+
+def test_cmpf_get():
+    a = Constant.from_float_and_width(1.0, f32)
+    b = Constant.from_float_and_width(2.0, f32)
+
+    cmpi_op = Cmpf.get(a, b, 1)
```
Comment (webmiche) [inline tests/dialects/test_arith.py:113]:
```suggestion
    cmpf_op = Cmpf.get(a, b, 1)
```

File: tests/dialects/test_arith.py [hunk 4]
```diff
@@ -82,3 +82,46 @@ def test_extend_truncate_fpops():
...
+    assert cmpi_op.rhs.typ == f32
+    assert cmpi_op.predicate.value.data == 1
+
+
+def test_cmpf_missmatch_type():
+    a = Constant.from_float_and_width(1.0, f32)
+    b = Constant.from_float_and_width(2.0, f64)
+
+    with pytest.raises(TypeError) as e:
+        cmpi_op = Cmpf.get(a, b, 1)
```
Comment (webmiche) [inline tests/dialects/test_arith.py:125]:
```suggestion
        cmpf_op = Cmpf.get(a, b, 1)
```

**Review by webmiche — APPROVED** [review]

File: tests/dialects/test_arith.py (hunk 4 above)
Comment (mesham) [reply tests/dialects/test_arith.py]: Good spot, done :)

File: tests/dialects/test_arith.py (hunk 3 above)
Comment (mesham) [reply tests/dialects/test_arith.py]: Good spot, done :)

File: xdsl/dialects/arith.py (hunk 1 above)
Comment (mesham) [reply xdsl/dialects/arith.py]: Good idea, I've done that

File: xdsl/dialects/arith.py [hunk 5]
```diff
@@ -412,13 +413,13 @@ def get(operand1: Union[Operation, SSAValue],
                            result_types=[operand1.typ])
 
 
-@irdl_op_definition
-class Cmpi(Operation):
+@dataclass
+class ComparisonOperation():
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:417]: :+1:

**Review by superlopuh — APPROVED** [review]

**Pull request merged**